### PR TITLE
Finer s3d stereo cost volume (num_bins 24→48) for sharper close-range depth

### PR DESCRIPTION
**Pull request by Bovey0809**
## What

Raise the s3d stereo cost-volume disparity sampling from **24 → 48 bins** over the unchanged `max_disp=48` (`StereoCostVolume` args in `yolo26-s3d.yaml`). At 24 bins over 0–48 feature-px the volume sampled only every ~2nd integer disparity; 48 bins samples **every integer disparity**, giving the depth branches a finer correlation profile. Depth is the dominant 3D-IoU error, so this directly attacks the AP3D ceiling (architecture lever from the depth-resolution plan).

## Result (cube_s3d, close-range stereo)

Per-component profiling over the 90-object val set (mean over 3 seeds, EMA weights, imgsz 768×1216) — AP3D on this tiny 5 cm-cube set is a single-epoch spike, so I compare on the stable continuous metrics the plan prescribes:

| metric | baseline (24 bins) | **48 bins** |
|---|---|---|
| depth error \|dz\| | 2.47 cm | **2.01 cm (−19%)** |
| frames with 3D IoU>0.5 | 14% | **30% (~2×)** |
| mean 3D IoU | 0.257 | **0.314** |

Improvement holds across all 3 seeds (larger than the seed-to-seed spread).

## KITTI guard

`max_disp` is unchanged, so this is strictly *more* disparity samples over the same range (no risk to KITTI's large disparities). Functional guard on `kitti-stereo8` (rect 384×1248, 40 ep): both configs train cleanly, no crash; the 48-bin run's stereo-disparity loss is actually lower (14.5 vs 21.1). A full-KITTI AP A/B can follow if desired.

## Cost

~2× cost-volume correlation channels at the P2/4 tap only (feeds the depth branches); negligible in the full model. 2D detection features are untouched.

Deleted: nothing. This is a one-value config change (num_bins 24→48) with no net line change and no code added — deletion/relocation don't apply. Reused the existing `StereoCostVolume` (no new module).

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Increased YOLO26-S3D stereo disparity sampling from 24 to 48 levels for finer depth estimation. 🎯

### 📊 Key Changes

- Updated `StereoCostVolume` in `yolo26-s3d.yaml` to use 48 disparity samples instead of 24.
- Maintains the same stereo feature output shape while reducing the approximate disparity step size.
- Leaves the rest of the detection head and PAN structure unchanged.

### 🎯 Purpose & Impact

- Improves depth and distance resolution by enabling finer stereo matching. 🔍
- May benefit 3D detection accuracy, especially for objects requiring precise depth estimates.
- Increases stereo cost-volume computation and memory usage, which could slightly reduce inference speed or require additional hardware resources. ⚙️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Bovey0809, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated welcome message, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/cfg/models/26/yolo26-s3d.yaml [hunk 1]
```diff
@@ -57,7 +57,7 @@ head:
   - [-1, 2, C3k2, [256, True]] # 16 (P3/8, clean — no cost vol)
 
   # Stereo cost volume (separate, NOT merged into P3)
-  - [1, 1, StereoCostVolume, [64, 48, 24]] # 17: [B, 64, H/8, W/8]
+  - [1, 1, StereoCostVolume, [64, 48, 48]] # 17: [B, 64, H/8, W/8] (48 disparity samples = every integer over max_disp; finer depth)
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/models/26/yolo26-s3d.yaml]: 💡 **MEDIUM**: `StereoCostVolume` generates offsets with `torch.linspace(0, max_disp, num_bins)` inclusively. With `max_disp=48` and `num_bins=48`, the rounded offsets contain 48 values across 0–48 and necessarily skip one integer (specifically 24); this does not provide a sample for every integer disparity in the configured range as the comment/PR states. Use 49 bins to cover all integer offsets 0 through 48, or adjust the stated range/expectation.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The configuration change is otherwise consistent with the cost-volume interface and preserves its output channel count and resolution. However, the requested every-integer disparity coverage has an inclusive-endpoint off-by-one error.

💬 Posted 1 inline comment

**Comment by Bovey0809** [comment]
Note: this is an architecture change — the cost-volume refine conv in-channels go 24→48, so **existing 24-bin s3d checkpoints cannot be fine-tuned into this config** (PyTorch will raise on strict load); retraining from the COCO-pretrained backbone is required. `Deleted: nothing` — hyperparameter/config tune, no code was removable.

**Comment by Bovey0809** [comment]
## Full-KITTI A/B (regression check) — cv48 slightly regresses KITTI

Ran cv24 vs cv48 on full kitti-stereo (3712/3769, DDP 4+4, 100 ep pretrained, rect 384×1248, default weights, identical recipe both arms). Car AP3D (R40):

| | @.5 Easy | @.5 Mod | @.5 Hard | @.7 Mod |
|---|---|---|---|---|
| cv24 | 0.292 | 0.218 | 0.183 | 0.0134 |
| cv48 | 0.267 | 0.202 | 0.173 | 0.0124 |

cv48 is consistently ~7–9% **lower** across all difficulties. So the finer cost volume is a **close-range-specific** win (large on the cube dataset) that does **not** transfer to KITTI's far-range regime and slightly regresses it.

Caveats: single seed, under-converged (auto→AdamW/100 ep, not the tuned KITTI MuSGD/200 ep — abs Mod@.5 21.8 vs ~34.9 documented), so the small margin is partly noise; but the sign is consistent across 6 cells.

**Recommendation: do not merge this as the shared default.** `num_bins=48` should be opt-in for close-range deployments (keep the general default at 24). Holding this PR pending that decision.

**Comment by Bovey0809** [comment]
Closing: the full-KITTI A/B (above) shows num_bins 24→48 slightly regresses KITTI (~7–9% Car AP3D), so it's not a valid change to the *shared* s3d default. It's a confirmed close-range win (cube_s3d: depth error −19%, 3D IoU>0.5 doubled) and stays in that deployment's own model config/checkpoint. Keeping the upstream default at 24. No duplicate config added.
